### PR TITLE
issues #3216 and #2512 - support for alternative reference schemes and bundle ref resolution from FHIRPath ResolveFunction

### PR DESCRIPTION
**Pull request by lmsurpre**
This PR adds support for alternative schemes, like `resource:`, in our
Reference type checking / reference value validation.
Previously, we required reference values to use the `http:`, `https:`, or `urn:` schemes,
but that prevented us from supporting alternative schemes like `resource:` which is
used in the SMART health cards spec.

I thought that I would need to do additional logic for supporting
reference rewriting for this scheme but, based on the tests I added in
FHIRRestHelperTest and ReferenceMappingVisitorTest, apparently not
...we just needed to let it past our parser.

I also included updates to our FHIRPath ResolveFunction implementation so that we can more easily
resolve local bundle references from FHIRPath.

Signed-off-by: Lee Surprenant <lmsurpre@us.ibm.com>

### CODE REVIEW

File: fhir-path/src/main/java/com/ibm/fhir/path/function/ResolveFunction.java [hunk 1]
```diff
@@ -94,16 +112,49 @@ public int getMaxArity() {
...
+                            // we know the root of the tree is a Bundle and the current node is of type Reference,
+                            // so walk up the tree until we get to highest Bundle.entry.fullUrl in the tree
+                            // and save the value of its fullUrl
+                            String fullUrl = null;
+                            FHIRPathTree tree = evaluationContext.getTree();
+                            FHIRPathNode nodeUnderTest = tree.getParent(node);
+                            while (nodeUnderTest != root) {
+                                if (nodeUnderTest.isResourceNode()) {
+                                    FHIRPathNode sibling = tree.getSibling(nodeUnderTest, "fullUrl");
+                                    fullUrl = sibling.getValue().asStringValue().string();
```
Comment (tbieste) [inline fhir-path/src/main/java/com/ibm/fhir/path/function/ResolveFunction.java:125]: Looks like null checks are done a lot, except here.  Just making sure this won't ever be null (or check for it, if not).

Comment (lmsurpre) [reply fhir-path/src/main/java/com/ibm/fhir/path/function/ResolveFunction.java:125]: honestly not sure if its possible for it to be null, but won't hurt to add some guards...will update

Comment (lmsurpre) [reply fhir-path/src/main/java/com/ibm/fhir/path/function/ResolveFunction.java]: done

File: fhir-path/src/test/java/com/ibm/fhir/path/test/ResolveInternalFragmentReferenceTest.java [hunk 2]
```diff
@@ -1,67 +0,0 @@
...
-                        .build())
-                    .build())
-                .build();
-
-        EvaluationContext evaluationContext = new EvaluationContext(bundle);
-        FHIRPathResourceNode resourceNode = evaluationContext.getTree().getNode("Bundle.entry[0].resource.contained[0]").asResourceNode();
-
-        Collection<FHIRPathNode> result = FHIRPathEvaluator.evaluator().evaluate(evaluationContext, "providedBy.exists() and providedBy.resolve().partOf.resolve().is(Organization)", resourceNode);
-
-        assertEquals(result, SINGLETON_TRUE);
```
Comment (prb112) [inline fhir-path/src/test/java/com/ibm/fhir/path/test/ResolveInternalFragmentReferenceTest.java:65]: What are the thoughts behind the test? too brittle? covered else where?

Comment (prb112) [reply fhir-path/src/test/java/com/ibm/fhir/path/test/ResolveInternalFragmentReferenceTest.java:65]: Found it

File: fhir-path/src/test/java/com/ibm/fhir/path/test/ResolveFunctionTest.java [hunk 3]
```diff
@@ -1,40 +1,249 @@
...
             result = evaluator.evaluate(evaluationContext, "Observation.subject.where(resolve() is Device)");
-            System.out.println("result: " + result);
-            System.out.println("");
+            assertNotNull(result);
+            assertEquals(result.size(), 0);
+        }
+    }
+
+    @Test
+    public void testResolveInternalFragmentReference() throws Exception {
```
Comment (prb112) [inline fhir-path/src/test/java/com/ibm/fhir/path/test/ResolveFunctionTest.java:64]: from another class...

File: fhir-path/src/main/java/com/ibm/fhir/path/function/ResolveFunction.java [hunk 4]
```diff
@@ -94,16 +112,51 @@ public int getMaxArity() {
...
                     } else {
-                        Matcher matcher = REFERENCE_PATTERN.matcher(referenceReference);
-                        if (matcher.matches()) {
-                            resourceType = matcher.group(RESOURCE_TYPE_GROUP);
-                            if (referenceType != null && !resourceType.equals(referenceType)) {
-                                throw new IllegalArgumentException("Resource type found in reference URL does not match reference type");
+                        if (isBundleContext) {
+                            // we know the root of the tree is a Bundle and the current node is of type Reference,
+                            // so walk up the tree until we get to highest Bundle.entry.fullUrl in the tree
+                            // and save the value of its fullUrl
```
Comment (prb112) [inline fhir-path/src/main/java/com/ibm/fhir/path/function/ResolveFunction.java:118]: While the comment really does help... .

Could you layout a small example? or link to a test example? 

If I have to maintain it, it'd help to laser focus on the 'aaah-ha' that's what it does.

File: fhir-path/src/main/java/com/ibm/fhir/path/function/ResolveFunction.java [hunk 5]
```diff
@@ -94,16 +112,51 @@ public int getMaxArity() {
...
+                                    }
+                                }
+                                nodeUnderTest = tree.getParent(nodeUnderTest);
                             }
-                            String baseUrl = matcher.group(BASE_URL_GROUP);
-                            if ((baseUrl == null ||  matchesServiceBaseUrl(baseUrl)) && evaluationContext.resolveRelativeReferences()) {
-                                // relative reference
-                                resource = resolveRelativeReference(evaluationContext, node, resourceType, matcher.group(LOGICAL_ID_GROUP), matcher.group(VERSION_ID_GROUP));
+
+                            String bundleReference = FHIRUtil.buildBundleReference(reference, fullUrl);
```
Comment (prb112) [inline fhir-path/src/main/java/com/ibm/fhir/path/function/ResolveFunction.java:132]: Note for others reviewing... when fuller is null, it'll return the referenceUriString = ref.getReference().getValue();

when this condition is true (ref.getReference() != null && ref.getReference().getValue() != null)

**Review by prb112 — APPROVED** [review]
Left comments for clarification. 

LGTM

File: fhir-path/src/main/java/com/ibm/fhir/path/function/ResolveFunction.java [hunk 6]
```diff
@@ -72,6 +78,17 @@ public int getMaxArity() {
...
+        boolean isBundleContext = root != null && root.type() == FHIRPathType.FHIR_BUNDLE;
+        Map<String, Resource> bundleResources = new HashMap<>();
+        if (isBundleContext) {
+            Bundle bundle = root.asResourceNode().resource().as(Bundle.class);
+            for (Entry e : bundle.getEntry()) {
+                if (e.getResource() != null && e.getFullUrl() != null && e.getFullUrl().hasValue()) {
+                    bundleResources.put(e.getFullUrl().getValue(), e.getResource());
+                }
+            }
+        }
```
[the remark this answers is not in the document]
Comment (lmsurpre) [reply fhir-path/src/main/java/com/ibm/fhir/path/function/ResolveFunction.java:91]: my assumption is that we set the `rootResource` to the outermost bundle (the one we're evaluating against).
i tried to account for the "bundle-in-a-bundle" case in the algorithm below by continuing to loop until we've reached the top of the tree.

[the remark this answers is not in the document]
Comment (JohnTimm) [reply fhir-path/src/main/java/com/ibm/fhir/path/function/ResolveFunction.java:91]: We have some helper functions (in FHIRPathUtil) that get the right resource depending on how it might be nested. These are typically stashed in the EvaluationContext.

File: fhir-path/src/main/java/com/ibm/fhir/path/function/ResolveFunction.java (hunk 4 above)
Comment (lmsurpre) [reply fhir-path/src/main/java/com/ibm/fhir/path/function/ResolveFunction.java]: I added some examples to this comment (and also a test case for the Bundle-in-Bundle case)

File: fhir-path/src/main/java/com/ibm/fhir/path/function/ResolveFunction.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (lmsurpre) [reply fhir-path/src/main/java/com/ibm/fhir/path/function/ResolveFunction.java:91]: I added the test.  I can think of a couple "interesting edge cases" but they seem so edge to me, that I don't care so much about what our behavior is there.

1. what to do if we have a Bundle thats contained within some other resource that is *not* a bundle (and that outer resource is the root/context).  With the current impl, we will not do local bundle resolution in that case, which seems fine to me.

2. what to do if we have a relative literal reference (e.g. `Patient/1`) in a Bundle.entry where that Bundle is inside another Bundle but the outer Bundle has no fullUrl for this entry.  in this scenario, I think the current algorithm will use the fullUrl from the inner bundle's entry to convert the relative URL into an absolute one.  I don't know if that behavior is desired or not, but again it seems like such an edge case that the impl seems ok to me

File: fhir-path/src/test/java/com/ibm/fhir/path/test/ResolveFunctionTest.java [hunk 7]
```diff
@@ -1,40 +1,293 @@
...
+import com.ibm.fhir.model.type.code.BundleType;
+import com.ibm.fhir.model.type.code.HTTPVerb;
 import com.ibm.fhir.path.FHIRPathNode;
+import com.ibm.fhir.path.FHIRPathResourceNode;
+import com.ibm.fhir.path.FHIRPathType;
 import com.ibm.fhir.path.evaluator.FHIRPathEvaluator;
 import com.ibm.fhir.path.evaluator.FHIRPathEvaluator.EvaluationContext;
+import com.ibm.fhir.path.util.FHIRPathUtil;
 
 public class ResolveFunctionTest {
```
Comment (punktilious) [inline fhir-path/src/test/java/com/ibm/fhir/path/test/ResolveFunctionTest.java:40]: javadoc

File: fhir-path/src/test/java/com/ibm/fhir/path/test/ResolveFunctionTest.java [hunk 8]
```diff
@@ -1,40 +1,293 @@
...
-    public static void main(String[] args) throws Exception {
+    @Test
+    public static void testResolve() throws Exception {
```
Comment (punktilious) [inline fhir-path/src/test/java/com/ibm/fhir/path/test/ResolveFunctionTest.java:42]: A comment block would be useful to describe what this is testing

File: fhir-path/src/test/java/com/ibm/fhir/path/test/ResolveFunctionTest.java [hunk 9]
```diff
@@ -1,40 +1,293 @@
...
+        FHIRPathResourceNode resourceNode = evaluationContext.getTree()
+                .getNode("Bundle.entry[0].resource.contained[0]").asResourceNode();
+
+        Collection<FHIRPathNode> result = FHIRPathEvaluator.evaluator().evaluate(evaluationContext,
+                "providedBy.exists() and providedBy.resolve().partOf.resolve().is(Organization)", resourceNode);
+
+        assertEquals(result, SINGLETON_TRUE);
+    }
+
+    @Test
```
Comment (punktilious) [inline fhir-path/src/test/java/com/ibm/fhir/path/test/ResolveFunctionTest.java:101]: comment block

File: fhir-path/src/test/java/com/ibm/fhir/path/test/ResolveFunctionTest.java (hunk 7 above)
Comment (lmsurpre) [reply fhir-path/src/test/java/com/ibm/fhir/path/test/ResolveFunctionTest.java:43]: added

File: fhir-path/src/test/java/com/ibm/fhir/path/test/ResolveFunctionTest.java (hunk 8 above)
Comment (lmsurpre) [reply fhir-path/src/test/java/com/ibm/fhir/path/test/ResolveFunctionTest.java:48]: done

**Review by punktilious — APPROVED** [review]
LGTM

File: fhir-model/src/test/java/com/ibm/fhir/model/util/test/ReferenceMappingVisitorTest.java [hunk 10]
```diff
@@ -193,11 +193,40 @@ public void testUpdateReferencesWithRelativeFullUrlAndRelativeReferenceMatch() t
         assertEquals(result.getGeneralPractitioner().get(0).getReference().getValue(), "Practitioner/1");
```
[the remark this answers is not in the document]
Comment (lmsurpre) [reply fhir-model/src/test/java/com/ibm/fhir/model/util/test/ReferenceMappingVisitorTest.java:193]: thanks Troy.  I'll
1. throw a ci-skip label on here to prevent the tests from running again
2. ammend my last commit to fix this
3. merge the PR

[the remark this answers is not in the document]
Comment (lmsurpre) [reply fhir-model/src/test/java/com/ibm/fhir/model/util/test/ReferenceMappingVisitorTest.java:193]: done
